### PR TITLE
Memory measurement and Compilation Time

### PR DESCRIPTION
**Pull request by anijain2305**
Code for https://github.com/pytorch/torchdynamo/issues/716
~~~
python benchmarks/huggingface.py --float32 -dcuda --training --use-eval-mode --profile-backend=pytorch --isolate --output=hf_pytorch.csv |& tee log1.txt
python benchmarks/huggingface.py --float32 -dcuda --training --use-eval-mode --profile-backend=eager --isolate --output=hf_eager.csv |& tee log2.txt
python benchmarks/huggingface.py --float32 -dcuda --training --use-eval-mode --profile-backend=aot_nop  --isolate --output=hf_aot_nop.csv |& tee log3.txt
python benchmarks/huggingface.py --float32 -dcuda --training --use-eval-mode --profile-backend=aot_nvfuser --isolate --output=hf_aot_nvfuser.csv |& tee log4.txt
python benchmarks/huggingface.py --float32 -dcuda --training --use-eval-mode --profile-backend=inductor_cudagraphs --isolate --output=hf_inductor_cudagraphs.csv |& tee log5.txt
~~~

### CODE REVIEW

File: benchmarks/common.py [hunk 1]
```diff
@@ -792,6 +792,71 @@ def decay_batch_exp(self, batch_size, factor=0.5, divisor=2):
             out_batch_size = batch_size - 1
         return max(0, int(out_batch_size))
 
+    def measure_latency_and_memory(self, device, model_name, model_iter_fn, backends):
```
Comment (voznesenskym) [inline benchmarks/common.py:795]: Minor, but we should make it clear in the API that this is GPU memory, it could easily get conflated/confused/added CPU memory.

File: benchmarks/common.py [hunk 2]
```diff
@@ -792,6 +792,71 @@ def decay_batch_exp(self, batch_size, factor=0.5, divisor=2):
...
+                    ctx = torchdynamo.optimize(backend)
+
+                # Reset and warmup
+                torchdynamo.reset()
+                torch.cuda.empty_cache()
+                t0 = time.perf_counter()
+                with ctx:
+                    model_iter_fn(model, example_inputs)
+                    model_iter_fn(model, example_inputs)
+                    model_iter_fn(model, example_inputs)
```
Comment (voznesenskym) [inline benchmarks/common.py:817]: Why do we run thrice between t0 and t1?

File: benchmarks/common.py [hunk 3]
```diff
@@ -792,6 +792,71 @@ def decay_batch_exp(self, batch_size, factor=0.5, divisor=2):
             out_batch_size = batch_size - 1
         return max(0, int(out_batch_size))
 
+    def measure_latency_and_memory(self, device, model_name, model_iter_fn, backends):
+        get_peak_memory = lambda: torch.cuda.max_memory_allocated() / 10**9
```
Comment (voznesenskym) [inline benchmarks/common.py:796]: Device is passed into `measure_latency_and_memory` but not `max_memory_allocated` which could lead to confusion. Is this intentional?

File: benchmarks/common.py [hunk 4]
```diff
@@ -792,6 +792,71 @@ def decay_batch_exp(self, batch_size, factor=0.5, divisor=2):
...
+                model_name,
+                self._args.training,
+                self._args.use_eval_mode,
+            )
+        except NotImplementedError:
+            logging.warn(f"{model_name} failed to load")
+
+        # Collect the compile and memory latency for backend
+        for backend in backends:
+            mem_experiment(backend)
```
Comment (voznesenskym) [inline benchmarks/common.py:846]: I don't know CUDA memory well, but retreating back to CPU memory / RAM, I would consider giving each backend its own process. What would happen here / how would our numbers look if a backend leaked?

File: benchmarks/common.py [hunk 5]
```diff
@@ -792,6 +792,71 @@ def decay_batch_exp(self, batch_size, factor=0.5, divisor=2):
...
+                    model_iter_fn(model, example_inputs)
+                peak_memory = get_peak_memory()
+            except:
+                latency = 0
+                peak_memory = 0
+
+            latencies.append(latency)
+            mems.append(peak_memory)
+
+        assert device == "cuda"
```
Comment (voznesenskym) [inline benchmarks/common.py:833]: nit: Move precondition way to the top of the func?

File: benchmarks/common.py (hunk 4 above)
Comment (anijain2305) [reply benchmarks/common.py:846]: Yeah, thats good point. Its relevant to GPU memory as well. Will run each backend in isolation.

File: benchmarks/common.py (hunk 3 above)
Comment (anijain2305) [reply benchmarks/common.py:796]: Nah, I am just bad at naming :)

Will try to name things better.

File: benchmarks/common.py (hunk 2 above)
Comment (anijain2305) [reply benchmarks/common.py:817]: This is just warmup. t0 and t1 are there to measure compilation cost. Torchscript can take a couple of iterations before it stop recompiling. 

This cost is approximation for compilation latency. It is approximation because it includes model iteration time as well. However, model iteration time typically is in order of ms. Therefore, t1 - t0 is only for flagging really long compilation times (>10 seconds).

**Comment by yushangdi** [comment]
the current code doesn't take `--batch-size-file` because it is run before the batch-size-file is processed. Maybe we want to  make `--peak-memory-for-backend` work with  `--batch-size-file` ?

**Comment by anijain2305** [comment]
> the current code doesn't take `--batch-size-file` because it is run before the batch-size-file is processed. Maybe we want to make `--peak-memory-for-backend` work with `--batch-size-file` ?

Good catch! Yes, thats correct. Do you want to make take over this PR? You can do `git fetch` and then `git checkout peak-mem` in your torchdynamo repo, and you will have access to this PR.

**Comment by anijain2305** [comment]
@voznesenskym @yushangdi Can you please review again?

**Review by voznesenskym — APPROVED** [review]
